Key-Manager Standard (#599)
SovereignCloudStack/standards: Standards/scs-XXXX-vN-key-manager-standard.md

---
title: Key-Manager Standard
type: Standard
status: Draft
track: IaaS
---

## Introduction

To encrypt user data like volumes or in the future also Images and ephemeral storage for VMs, the key is needed to be present in the infrastructure.
Therefore a key-manager is needed to store the keys and apply authorization policies on requests.
The OpenStack key-manager implemenation that can bu utilized for this is named Barbican.
This standard aims to provide a base level of security for Cloud Service Providers that integrate a key-manager into their deployments.

## Terminology

| Term | Meaning |
|---|---|
| API | Application Programming Interface, often referring to the REST API interfaces provided by OpenStack and related services |
| CSP | Cloud Service Provider, provider managing the OpenStack infrastructure |
| IaaS | Infrastructure-as-a-Service |
| HSM | Hardware Security Module |

> | KEK | Key Encryption Key |

Reviewer: I suggest to add Barbican into the Terminology to reduce repeating in the spec itself.

## Motivation

When user data is encrypted the keys need to be stored securely in the Infrastructure to be present, when a possible encryption or decryption needs to take place.
Key-managers like Barbican provide this functionality on the IaaS-Level.
Using such a key-manger and thus provide to customers the possibility to encrypt user data is not yet implemented everywhere in IaaS deployments.
A first step towards more security is to encourage CSPs to use this service and provide the feature of using encrypted resources to customers.
It is also important to take a closer look into the key-manager and to apply aim for an appropiate level of security there.

> The Key-Manager is responsible for storing the keys securely, which can be done by encrypting them with a KEK and storing them in a database and store the KEK either also encrypted in the database or to store it somewhere else.

Reviewer: The Key-Manager service manages keys in a secure manner. This can be achieved differently and is not primarily in scope of this standard.

> Barbican as the OpenStack implementation of a key-manager is relying on a Master-KEK, which encrypts project-specific KEKs, which encrypt the Keys within a project.

Reviewer: Barbican stores keys encrypted with the project specific KEK, including the KEK itself, in the database. Master KEK, used to encrypt project specific KEK is not stored in the database and is stored differently depending on the backend storage plugin used.

All keys except for the Master-KEK are stored in a database.
This Master-KEK is stored differently for each plugin and needs to be protected.

## Design Considerations

While discussing what this standard should aim for it was discovered that some CSPs don't use Barbican or another key-manager at all and do not provide the feature to encrypt user data to their customers.
This should change, but the exact change comes with financial burden, when choosing a plugin in Barbican to store the Master-KEK.
To minimize the burden and enable more CSPs to step up and provide encryption, this standard will only make recommendations about plugins.

### Options considered

#### _Option 1_

It was considered to only recommend a certain set of plugins or backends for the key-manager, but this may be very prone to change if Barbican adds a new plugin.
As the SCS only wants to mandate the API that can be abstracted through the Castellan library in OpenStack, integrating any other key-manager implementation is not uncommon, so this standard needs to consider other possible key-managers as well.
Due to these reasons this Option was disregarded.

#### _Option 2_

> Looking into the plugins and possible attack vectors one design decision in the plugins is very important: where and how to store the Master-KEK.

Reviewer: Looking at the available Barbican plugins ...

---
title: Key-Manager Standard
type: Standard
status: Draft
track: IaaS
---

## Introduction

To encrypt user data like volumes or in the future also Images and ephemeral storage for VMs, the key is needed to be present in the infrastructure.

> Therefore a key-manager is needed to store the keys and apply autorization policies on requests.

Reviewer 2: typo:
```suggestion
Therefore a key-manager is needed to store the keys and apply authorization policies on requests.
```

The OpenStack key-manager implemenation that can bu utilized for this is named Barbican.
This standard aims to provide a base level of security for Cloud Service Providers that integrate a key-manager into their deployments.

## Terminology

| Term | Meaning |
|---|---|
| API | Application Programming Interface, often referring to the REST API interfaces provided by OpenStack and related services |
| CSP | Cloud Service Provider, provider managing the OpenStack infrastructure |
| IaaS | Infrastructure-as-a-Service |
| HSM | Hardware Security Module |
| KEK | Key Encryption Key |

## Motivation

When user data is encrypted the keys need to be stored securely in the Infrastructure to be present, when a possible encryption or decryption needs to take place.

> Key-managers like Barbican provide this functionality on IaaS-Level.

Reviewer 2: nit:
```suggestion
Key-managers like Barbican provide this functionality on the IaaS-Level.
```

The Key-Manager is responsible for storing the keys securely, which can be done by encrypting them with a KEK and storing them in a database and store the KEK either also encrypted in the database or to store it somewhere else.
Barbican as the OpenStack implementation of a key-manager is relying on a Master-KEK, which encrypts project-specific KEKs, which encrypt the Keys within a project.
All keys except for the Master-KEK are stored in a database.
This Master-KEK is stored differently for each plugin and needs to be protected.

## Design Considerations

While discussing what this standard should aim for it was discovered that some CSPs don't use Barbican or another key-manager at all and do not provide the feature to encrypt user data to their customers.
This should change, but the exact change comes with financial burden, when choosing a plugin in Barbican to store the Master-KEK.
To minimize the burden and enable more CSPs to step up and provide encryption, this standard will only make recommendations about plugins.

### Options considered

#### _Option 1_

> Barbican as the OpenStack key-manager implementation offers different plugins to store the Master-KEK.

Reviewer 2: I feel this sentence is a little bit redundant with the section in lines 32-35, maybe it can be removed?

Author: I removed it and rephrased the whole paragrph a little bit

It was considered to only recommend a certain set of plugins, but this may be very prone to change if Barbican adds a new plugin and this standard would need to consider other possible key-managers as well.
As the SCS only wants to mandate the API that can be abstracted through the Castellan library in OpenStack, integrating any other key-manager implementation is not uncommon.
So this Option was disregarded.

#### _Option 2_

Looking into the plugins and possible attack vectors one design decision in the plugins is very important: where and how to store the Master-KEK.
Because the Plugins might use different technologies, but most of them increase the security level by not storing the Master-KEK in plain text on the physical machine Barbican is running on.
This mechanism as a whole, is something that CSPs should aim to do.

## Standard

To increase the level of security and overall user data encryption CSPs SHOULD implement the key-manager API (e.g. implemented by barbican) with a security level of storing Keys encrypted and storing the KEK in another place than the Keys.

> If possible CSPs SHOULD NOT store the Master-KEK in plain-text on the physical host the key-manager is running on.

Reviewer 2: I think it would be helpful here to explicitly call out the Barbican plugins which do store keys unencrypted on local disc, so CSPs can avoid these, or do the reverse and at least list some plugins which are deemed "secure" (enough).

Author: The problem here is:
the simple crypto plugin does store the masterKEK unencrypted in the Barbican config. So we can name it as a negative example. But I do not think we should deem this plugin as completely insecure.
E.g. it can be managed to let Barbican run in enclaves like SGX, to protect the config.
And having Barbican with simpley crypto is still better than not having any option to encrypt user data at all - so I am a bit hesitant to give a "not do this"-example.

We could add a sentence stating, that using HSMs would be one option to fulfill this, but this could give CSPs the impression, that we will put a high financial burden on them.

I am skeptical of both options and would like to have more input from other reviewers on this

Reviewer: I would add another sentence like: "Barbican supports deploying out-of-tree drivers what enables operators to satisfy their specific needs." With this sentence I would like to give a signal to the CSPs that they have quite lot possibilities in their hands to offer better security to their customers.

Author: @gtema this is more of an implementation note and not part of the standard. Should I still add this somewhere in this document or just reference the implementation notes somewhere?

Maybe an:
```
:::tip

Barbican supports deploying out-of-tree drivers what enables operators to satisfy their specific needs.

:::
```
should be added?

Reviewer: yupp, adding a tip sound good (I agree that we can place it in the impl. part). I just want to ensure, that CSPs are not frightened by the standard and know that it is possible to have some custom stuff running what would satisfy the standard.

Reviewer 2: from my pov this conversation is resolved, @gtema ?

Reviewer 3: You need barbican anyway alone as a requirement for octavia if you do ssl offloading with amphoras too. Also other Keymanager appliances plugin into "barbican" as drivers/plugins

---
title: Key-Manager Standard
type: Standard
status: Draft
track: IaaS
---

## Introduction

> To encrypt user data like volumes or in the future also Images and ephemeral storage for VMs, the key is need to be known in the infrastructure.

Reviewer: "the key is need" is syntactically wrong

> To provide the key to those operations without including the user every time a Key Manager within the infrastructure can be utilized to store the keys and apply authorization policies on requests.

Reviewer: I needed to re-read this sentence 3 times before I understood the semantics. Please simplify

OpenStack offers a Key Manager implementation that is named Barbican, which provides these features.
This standard aims to provide a base level of security for Cloud Service Providers that integrate a Key Manager into their deployments.

## Terminology

| Term | Meaning |
|---|---|
| API | Application Programming Interface, often referring to the REST API interfaces provided by OpenStack and related services |
| Barbican | The Key Manager implementation in OpenStack |
| CSP | Cloud Service Provider, provider managing the OpenStack infrastructure |
| IaaS | Infrastructure-as-a-Service |
| HSM | Hardware Security Module |
| KEK | Key Encryption Key |

## Motivation

User data encryption requires an encryption key to be known during encryption and decryption processes.
Key Managers like Barbican provide this functionality on the IaaS-Level.
Not every IaaS deployment currently offers user data encryption as part of their standard offering.

> A first step towards more security is to encourage CSPs to provide better data security by offering data encryption to the customers.

Reviewer: "[snip] step towards more security is to encourage [snip] to provide better data security [snip]". The sentence is really awkward.

> It is also important to take a closer look into the Key Manager and to apply aim for an appropiate level of security there.

Reviewer: "apply aim" is not a proper word combination

## Design Considerations

While discussing what this standard should aim for it was discovered that some CSPs don't use Barbican or another Key Manager at all and do not provide the feature to encrypt user data to their customers.
This should change, but the exact change comes with financial burden, when choosing a plugin in Barbican to store the Master-KEK.
To minimize the burden and enable more CSPs to step up and provide encryption, this standard will only make recommendations about plugins.

### Options considered

#### _Option 1_

It was considered to only recommend a certain set of plugins or backends for the Key Manager, but this may be very prone to change if e.g. Barbican adds a new plugin.
As the SCS only wants to mandate the API that can be abstracted through the Castellan library in OpenStack, integrating any other Key Manager implementation is not uncommon, so this standard needs to consider other possible Key Managers as well.
Due to these reasons this option was disregarded.

#### _Option 2_

Looking into the available Barbican plugins and possible attack vectors one design decision in the plugins is very important: where and how to store the Master-KEK.

> Because the Plugins might use different technologies, but most of them increase the security level by not storing the Master-KEK in plain text on the physical machine Barbican is running on.

Reviewer: "Because [A], but [B]." - I can't understand the sentence.

This mechanism as a whole, is something that CSPs should aim to do.

## Standard

> To increase the level of security and overall user data encryption CSPs SHOULD implement the Key Manager API (e.g. implemented by Barbican) with a security level of storing Keys encrypted and storing the Master-KEK in another place than the Keys.

Reviewer: Please rephrase the sentence. I do not think that KeyManager API has any "security level" with not storing key encrypted.

Author: I broke it up into two sentences.

## Introduction

To encrypt user data like volumes or in the future also Images and ephemeral storage for VMs, the key is need to be known in the infrastructure.

> To provide the key to those operations without includign the user every time a key-manager within the infrastructure can be utilized to store the keys and apply authorization policies on requests.

Reviewer 2: ```suggestion
To provide the key to those operations without including the user every time a key-manager within the infrastructure can be utilized to store the keys and apply authorization policies on requests.
```

OpenStack offers a key-manager implementation that is named Barbican, which provides these features.
This standard aims to provide a base level of security for Cloud Service Providers that integrate a key-manager into their deployments.

## Terminology

| Term | Meaning |
|---|---|
| API | Application Programming Interface, often referring to the REST API interfaces provided by OpenStack and related services |
| Barbican | The key-manager implementation in OpenStack |
| CSP | Cloud Service Provider, provider managing the OpenStack infrastructure |
| IaaS | Infrastructure-as-a-Service |
| HSM | Hardware Security Module |
| KEK | Key Encryption Key |

## Motivation

User data encryption requires an encryption key to be known during encryption and decryption processes.
Key-managers like Barbican provide this functionality on the IaaS-Level.
Not every IaaS deployment currently offers user data encryption as part of their standard offering.

> A first step towards more security is to encourage CSPs to provide a better data security by offering data encryption to the customers.

Reviewer 2: nit:
```suggestion
A first step towards more security is to encourage CSPs to provide better data security by offering data encryption to the customers.
```

It is also important to take a closer look into the key-manager and to apply aim for an appropiate level of security there.
The Key-Manager service manages keys in a secure manner.
This can be achieved differently and is not primarily in scope of this standard.
Barbican stores keys encrypted with the project specific KEK, including the KEK itself, in the database.
The Master KEK, used to encrypt the project specific KEKs is not stored in the database and is stored differently depending on the backend storage plugin used.

> This standard also abstracts from used plugins and want to ensure that the Master-KEK is protected, too.

Reviewer 2: nit:
```suggestion
This standard also abstracts the used plugins and wants to ensure that the Master-KEK is protected, too.
```

#### _Option 1_

It was considered to only recommend a certain set of plugins or backends for the key-manager, but this may be very prone to change if Barbican adds a new plugin.
As the SCS only wants to mandate the API that can be abstracted through the Castellan library in OpenStack, integrating any other key-manager implementation is not uncommon, so this standard needs to consider other possible key-managers as well.

> Due to these reasons this Option was disregarded.

Reviewer 2: nit:
```suggestion
Due to these reasons this option was disregarded.
```
